Fix crash when Land activity runs on an already-landed aircraft.

pchote opened the pull request:
Fixes #21987.

PunkPun commented:
If I were to follow the reasoning in https://github.com/OpenRA/OpenRA/issues/21987#issuecomment-3138045587, then isn't the issue that influence doesn't get removed? This PR doesn't change that

PunkPun commented:
This crash is there to trigger activities not cleaning up their influence and this PR seems to only hide the error?

pchote commented:
The crash happens because `Land` is running on an actor that is already on the ground, whereas `Land` assumes that it will only ever be running on an airbourne actor. The testcase in #21987 demonstrates this trivially.

michaeldgg2 commented:
Pun's right, in our case (in OpenE2140), we had several nasty crashes related to our custom flying harvesters, all of them due to incorrect handling of influence. Aircraft are delicate actors, when it comes to their influence on terrain and other actors.

But this PR fixes different bug.

PunkPun reviewed (commented):
I see, the fix still doesn't make sense though. The already being on the ground case should be handled explicitly

pchote commented:
The issue occurs because the target land position generally has height 0 which doesnt match the land altitude. This case is handled correctly after the problematic if block.

PunkPun commented:
What if our influence is wrong and needs to be moved? Etc, etc.

pchote commented:
it already explicitly handles that case! The issue is just that it incorrectly assumes that "not at the height i've asked for" implies "still in the air".

pchote commented:
Updated. The first commit now contains the actual fix for the underlying bug. The second commit retains the original fix as it is still the correct thing it should be doing.

pchote commented:
It appears that there are multiple distinct bugs here - the first commit fixes the underlying cause of #21987. #21973 is resolved by the second commit, but that will be hiding the underlying cause.

pchote commented:
Indeed #21973 was distinct, and has already been fixed. I have dropped the second commit as I now see the value of not hiding the crash when we have inconsistent state.

pchote committed:
Fix landing altitude checks.

Both GetPosition and LandPosition were trying to solve the problem
of dynamic altitude offsets for carryalls with different-sized cargo.

However, because they both apply the offsets, when the two are used
together the two sets of offsets cancel to 0, producing incorrect behaviour.

GetPosition is removed in favor of using CenterPosition with the dynamic
LandAltitude. Fly.Tick and Carryall ignore the z coordinate so their
behaviour is unchanged. Land and Aircraft.AtLandAltitude now return
the correct result.

PunkPun reviewed (commented):
You can now trigger the crash by issuing an order for caryall to land in the same spot it is landed to

- In TS, build a caryall
- Pick up a vehicle
- Force land it to a location
- Force land it again

pchote commented:
Updated to fix the carryall crash. See commit message for detailed explanation.

PunkPun reviewed (approved):
Fixes several issues

PunkPun commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/619a405b74effd9707e82ad1b86ee698743588fd...a4c4aff5f1db88bd31dfc6f4948ca7905b765f0f
